enhancements/operator-conditions-lib.md

```diff
@@ -49,40 +49,111 @@ Operators need to be able to read the Status CR created by OLM, and also have th
+  // OperatorCondition to the provided status. If the condition is not
+  // present, it is added to the CR.
+  // To set a new condition, the user can call this method and provide optional
+  // parameters if required. It returns an error if there are problems getting or
+  // updating the OperatorCondition object.
+  Set(context.Context, metav1.ConditionType, metav1.ConditionStatus, ...Option) error
+}
+```
 
-1. Get the `NamespacedName` of the CRD:
+Optional fields like `Reason` or `Message` of the condition, can be set by passing the values to `Options`.
 
-In order to get the `Status` CRD from cluster using controller-runtimes's [client][cr_client], users will need [`ObjectKey`][cr_objectkey] of type [`types.NamespacedName`][cr_namespacedname]. OLM will set the name of Status CR as an [environment variable][conditions_followup_pr] while creating the resource. Namespace of the operator can be found from the associated [service account secret][accessing_api_from_pod] available in the operators' pod. The library will utilize both of these to obtain the `NamespacedName` for the CRD.
+```go
+// Option is a function that applies a change to a condition.
+// This can be used to set optional condition fields, like reasons
+// and messages.
+type Option func(*metav1.Condition)
+
+func WithReason(reason string) Option {
+	return func(c *metav1.Condition) {
+		c.Reason = reason
+	}
+}
+
+func WithMessage(message string) Option {
+	return func(c *metav1.Condition) {
+		c.Message = message
+	}
+}
+```
 
-2. Specify conditions in the CRD:
+The library will provide helpers to return a new `Conditions` interface when a controller-runtime client and the `conditionType` is provided.
 
-The operators can introduce OLM-supported conditions which influence OLM's behavior or can also specify any custom ([non-OLM supported][non-olm-supported-condition]) condition for communicating the status of the operator to users. An example Status CR is given below:
+```go
 
-```yaml
-apiVersion: operators.coreos.com/v1
-kind: Status
-metadata:
-  name: foo-operator-zdgen-asdf # RandomGen Name
-  namespace: operators
-status:
-  conditions:
-    type: Upgradeable # The name of the `foo` OLM Supported Condition.
-    status: "False"   # The operator is not ready to be upgraded.
-    reason: "migration"
-    message: "The operator is performing a migration."
-    lastTransitionTime: "2020-08-24T23:15:55Z"
+// conditionAccessor contains the controller-runtime client and the namespacedName
+// required to access the objects on cluster.
+type conditionAccessor struct {
+  client Client.client
+  namespacedName types.NamespacedName
+}
+
+// condition contains the conditionAccessor and the conditionType which
+// needs to be added/modified.
+type condition struct {
+  conditions conditionAccessor
+  conditionType metav1.ConditionType
+}
+
+// condition will implement the Set and Get methods defined in Condition interface.
+var _ Condition = &condition{}
```

Comment:
Is this supposed to be
```suggestion
var _ Conditions = &condition{}
```

I see both `Condition` and `Conditions` referenced in these changes. Are they the same thing?

Reply:
Have updated the PR. This was supposed to be `Conditions` interface.

Reply:
Since we're creating a "client" for a single condition defined by a `metav1.ConditionType`, wouldn't it make more sense to have `type Condition interface`?

Reply:
Yeah, I agree. Updated the EP